### ISSUE TITLE
Car travel game hook glitches

### ISSUE THREAD

**Issue by NovaRain** [phobos2077/sfall#220]
With a simple cheat script that disables fuel consumption for the car:
```
procedure start begin
   if game_loaded then begin
      register_hook(HOOK_CARTRAVEL);
   end else begin
      set_sfall_return(-1);
      set_sfall_return(0);
   end
end
```
The car body will disappear after entering a new map (random encounter or another location), leaving only the car trunk. Exiting and re-entering the map makes the car body appear.

And with a script that only pass and return the arguments:
```
procedure start begin
   if game_loaded then begin
      register_hook(HOOK_CARTRAVEL);
   end else begin
      variable
         speed := get_sfall_arg,
         fuel := get_sfall_arg;
      set_sfall_return(speed);
      set_sfall_return(fuel);
   end
end
```
The whole car (body + trunk) disappears after entering a new map and can't be made reappeared, as if the car is lost forever. In debug.log:
```
trunk: map_enter_p_proc
   Trunk_In_Party: start
   GVAR_CAR_PLACED_TILE == 5
  trying to hide trunk
   Trunk_In_Party: end
trunk: map_update_p_proc
   Trunk_In_Party: start
   GVAR_CAR_PLACED_TILE == 5
  trying to hide trunk
   Trunk_In_Party: end
```

Everything is normal if setting InjectAllGameHooks=1 in ddraw.ini, and it didn't happen in 4.1.2. I think it's related to the change for injecting HOOK_SETGLOBALVAR, as reverting the code change seems fixing this issue.

**Comment by FakelsHub** [comment]
Does the existence of a hook(Car travel) script in game cause this problem?

**Comment by NovaRain** [comment]
>Does the existence of a hook(Car travel) script in game cause this problem?

No, I'm sure there's only one script using the hook.

**Comment by FakelsHub** [comment]
I do not understand anything
On which version of sfall a glitch occurs? (current 4.1.3)
having an empty HOOK_SETGLOBALVAR hook script, fixes the problem?

**Comment by NovaRain** [comment]
>On which version of sfall a glitch occurs? (current 4.1.3)
having an empty HOOK_SETGLOBALVAR hook script, fixes the problem?

The builds are from the newest commit (both yours and here). And yes, with an empty HOOK_SETGLOBALVAR hook script it works properly.

>Today I traveled by car, it did not disappear.
so the question is because of whom does this bug?

Are you sure you didn't inject all hooks in ddraw.ini?

**Comment by FakelsHub** [comment]
No.

This is some kind of nonsense, how is the car connected and the hook, the script logic of the car gets the data from the hook?
okay i'll come home i'll check everything.

**Comment by FakelsHub** [comment]
https://github.com/phobos2077/sfall/blob/0a8c5d704c78168c1b69984669c20c0509ba71e5/sfall/Modules/HookScripts/MiscHs.cpp#L286

The reason for this is that the option of informing a change in karma is apparently enabled, because of this, cRet contains a "garbage" value.

**Comment by NovaRain** [comment]
>The reason for this is that the option of informing a change in karma is apparently enabled, because of this, cRet contains a "garbage" value.

Oh, you're right. The scripts work properly with DisplayKarmaChanges=0.

**Comment by FakelsHub** [comment]
This cRet worries me a lot, because as one is used for all the hooks, it was necessary correctly to make a separate cRet for each hook.
I could do it but I really need to redo all the hooks, I don’t want to do.
Now I suffer from initially incorrectly implemented hooks from TimeSlips

**Comment by FakelsHub** [comment]
Without injection, the same can happen if the execution of the hooks follows this way:
`hook1 start > ret from hook1 > hook2 start > ret from hook2 > hook2 end > check cRet = OK > hook1 end > check cRet = BUG`

**Comment by FakelsHub** [comment]
I think I need to somehow prohibit the use of set_global_var in its hook, because it will cause the game to hang. try experimenting with these two hooks without my last commit.
```
register_hook_proc(HOOK_COMBATDAMAGE, hook);
register_hook_proc(HOOK_SETGLOBALVAR,  hook1);

procedure hook begin
   display_msg("HOOK_COMBATDAMAGE");
   set_global_var(0, 100);
   set_sfall_return(1);
end
procedure hook1 begin
   display_msg("HOOK_SETGLOBALVAR");
   set_sfall_return(50);
end
```

**Comment by NovaRain** [comment]
I tested your script with the build from the repo here (without your last commit, obviously):
It displays two HOOK_SETGLOBALVAR after loading a save. I attack Vic and it displays HOOK_COMBATDAMAGE and HOOK_SETGLOBALVAR, dealing 50 points of damage to him and 1 point to myself, and my karma is dropped to 50. If I killed someone it will spam more than a dozen of HOOK_SETGLOBALVAR in the message box, but the game doesn't hang.
